# Data Exploration

A simple exploration of the train and test parquet files.

In [1]:
import pandas as pd

# Load the data
train = pd.read_parquet('train.parquet')
test = pd.read_parquet('test.parquet')

## Dataset Shapes

In [2]:
print(f"Train: {train.shape[0]:,} rows, {train.shape[1]} columns")
print(f"Test: {test.shape[0]:,} rows, {test.shape[1]} columns")

Train: 17,499,636 rows, 19 columns
Test: 4,393,179 rows, 19 columns


## Column Overview

In [3]:
train.dtypes

status                    int64
gender                   object
firstName                object
level                    object
lastName                 object
userId                   object
ts                        int64
auth                     object
page                     object
sessionId                 int64
location                 object
itemInSession             int64
userAgent                object
method                   object
length                  float64
song                     object
artist                   object
time             datetime64[us]
registration     datetime64[us]
dtype: object

## Sample Data

In [4]:
train.head(10)

status gender firstName level lastName   userId             ts  \
0        200      M     Shlok  paid  Johnson  1749042  1538352001000   
992      200      M     Shlok  paid  Johnson  1749042  1538352525000   
1360     200      M     Shlok  paid  Johnson  1749042  1538352703000   
1825     200      M     Shlok  paid  Johnson  1749042  1538352935000   
2366     200      M     Shlok  paid  Johnson  1749042  1538353200000   
3271     200      M     Shlok  paid  Johnson  1749042  1538353671000   
3802     200      M     Shlok  paid  Johnson  1749042  1538353937000   
6585     200      M     Shlok  paid  Johnson  1749042  1538355337000   
6675     200      M     Shlok  paid  Johnson  1749042  1538355388000   
6961     200      M     Shlok  paid  Johnson  1749042  1538355523000   

           auth       page  sessionId                         location  \
0     Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
992   Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
1360  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
1825  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
2366  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
3271  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
3802  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
6585  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   
6675  Logged In  Downgrade      22683  Dallas-Fort Worth-Arlington, TX   
6961  Logged In   NextSong      22683  Dallas-Fort Worth-Arlington, TX   

      itemInSession                                          userAgent method  \
0               278  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
992             279  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
1360            280  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
1825            281  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
2366            282  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
3271            283  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
3802            284  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
6585            285  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   
6675            286  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    GET   
6961            287  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...    PUT   

          length                                    song  \
0      524.32934  Ich mache einen Spiegel - Dream Part 4   
992    178.02404                 Monster (Album Version)   
1360   232.61995                       Seven Nation Army   
1825   265.50812        Under The Bridge (Album Version)   
2366   471.69261                            Circlesong 6   
3271   266.86649                         Who Can Compare   
3802  1400.65914                              Angel Dust   
6585   186.98404                         Sweet And Dandy   
6675         NaN                                    None   
6961   306.05016                             On The Moon   

                     artist                time        registration  
0                 Popol Vuh 2018-10-01 00:00:01 2018-08-08 13:22:21  
992                 Skillet 2018-10-01 00:08:45 2018-08-08 13:22:21  
1360      The White Stripes 2018-10-01 00:11:43 2018-08-08 13:22:21  
1825  Red Hot Chili Peppers 2018-10-01 00:15:35 2018-08-08 13:22:21  
2366         Bobby McFerrin 2018-10-01 00:20:00 2018-08-08 13:22:21  
3271         Foolish Things 2018-10-01 00:27:51 2018-08-08 13:22:21  
3802        Gil Scott Heron 2018-10-01 00:32:17 2018-08-08 13:22:21  
6585    Toots & The Maytals 2018-10-01 00:55:37 2018-08-08 13:22:21  
6675                   None 2018-10-01 00:56:28 2018-08-08 13:22:21  
6961         Peter Cincotti 2018-10-01 00:58:43 2018-08-08 13:22:21

## Missing Values

In [5]:
# Missing values in train
missing_train = train.isnull().sum()
missing_train[missing_train > 0]

length    3208203
song      3208203
artist    3208203
dtype: int64

In [6]:
# Missing values in test
missing_test = test.isnull().sum()
missing_test[missing_test > 0]

gender        653681
firstName     653681
lastName      653681
location      653681
userAgent     653681
length       1331368
song         1331368
artist       1331368
dtype: int64

## Basic Statistics

In [7]:
train.describe()

status            ts     sessionId  itemInSession        length  \
count  1.749964e+07  1.749964e+07  1.749964e+07   1.749964e+07  1.429143e+07   
mean   2.091387e+02  1.540428e+12  8.480294e+04   1.055937e+02  2.487135e+02   
min    2.000000e+02  1.538352e+12  1.000000e+00   0.000000e+00  5.220000e-01   
25%    2.000000e+02  1.539340e+12  2.515900e+04   2.600000e+01  1.998885e+02   
50%    2.000000e+02  1.540397e+12  7.903800e+04   6.600000e+01  2.340828e+02   
75%    2.000000e+02  1.541500e+12  1.383680e+05   1.440000e+02  2.768714e+02   
max    4.040000e+02  1.542672e+12  2.070030e+05   1.426000e+03  3.024666e+03   
std    3.023050e+01  1.233485e+09  6.141427e+04   1.168854e+02  9.722845e+01   

                             time                registration  
count                    17499636                    17499636  
mean   2018-10-25 00:47:01.161927  2018-08-25 04:40:21.543066  
min           2018-10-01 00:00:01         2017-10-14 22:05:25  
25%    2018-10-12 10:33:57.750000         2018-08-10 21:14:59  
50%           2018-10-24 15:58:54         2018-09-05 18:35:50  
75%           2018-11-06 10:25:35         2018-09-20 17:24:57  
max           2018-11-20 00:00:00         2018-11-19 23:34:34  
std                           NaN                         NaN

## Categorical Columns

In [8]:
# Page types
train['page'].value_counts()

page
NextSong                     14291433
Thumbs Up                      789391
Home                           645259
Add to Playlist                409606
Roll Advert                    284837
Add Friend                     262147
Logout                         204700
Thumbs Down                    164964
Downgrade                      124248
Settings                       101191
Help                            89035
Upgrade                         37696
About                           33117
Save Settings                   20370
Error                           17294
Submit Upgrade                  11381
Submit Downgrade                 4425
Cancellation Confirmation        4271
Cancel                           4271
Name: count, dtype: int64

In [9]:
# User levels
train['level'].value_counts()

level
paid    13506659
free     3992977
Name: count, dtype: int64

In [10]:
# Gender distribution
train['gender'].value_counts()

gender
M    9191364
F    8308272
Name: count, dtype: int64

## Unique Counts

In [11]:
print(f"Unique users (train): {train['userId'].nunique():,}")
print(f"Unique users (test): {test['userId'].nunique():,}")
print(f"Unique songs (train): {train['song'].nunique():,}")
print(f"Unique artists (train): {train['artist'].nunique():,}")
print(f"Unique sessions (train): {train['sessionId'].nunique():,}")

Unique users (train): 19,140
Unique users (test): 2,904
Unique songs (train): 239,299
Unique artists (train): 37,264
Unique sessions (train): 161,194


## Time Range

In [12]:
print(f"Train time range: {train['time'].min()} to {train['time'].max()}")
print(f"Test time range: {test['time'].min()} to {test['time'].max()}")

Train time range: 2018-10-01 00:00:01 to 2018-11-20 00:00:00
Test time range: 2018-10-01 00:00:06 to 2018-11-20 00:00:00


---

# Key Insights for Churn Prediction

The following analysis reveals critical patterns for predicting user churn.

## Churn Definition & Rate

A user has **churned** when they visit the `Cancellation Confirmation` page.

In [13]:
# Identify churned users
churned_users = train[train['page'] == 'Cancellation Confirmation']['userId'].unique()
all_users = train['userId'].unique()

n_churned = len(churned_users)
n_total = len(all_users)
churn_rate = n_churned / n_total

print(f"Total users in training set: {n_total:,}")
print(f"Churned users: {n_churned:,}")
print(f"Non-churned users: {n_total - n_churned:,}")
print(f"\nChurn rate: {churn_rate:.2%}")

Total users in training set: 19,140
Churned users: 4,271
Non-churned users: 14,869

Churn rate: 22.31%


## Activity Levels: Churned vs Non-Churned

Comparing overall event counts between churned and non-churned users.

In [14]:
# Compare activity levels: churned vs non-churned users
churned_set = set(churned_users)

# Calculate events per user
events_per_user = train.groupby('userId').size()

# Split by churn status
churned_events = events_per_user[events_per_user.index.isin(churned_set)]
non_churned_events = events_per_user[~events_per_user.index.isin(churned_set)]

print("Events per user:")
print(f"  Churned users:     mean = {churned_events.mean():.1f}, median = {churned_events.median():.1f}")
print(f"  Non-churned users: mean = {non_churned_events.mean():.1f}, median = {non_churned_events.median():.1f}")
print(f"\n-> Activity levels are similar between churned and non-churned users.")
print("   Churn does not appear to be driven by low engagement.")

Events per user:
  Churned users:     mean = 917.8, median = 557.0
  Non-churned users: mean = 913.3, median = 531.0

-> Activity levels are similar between churned and non-churned users.
   Churn does not appear to be driven by low engagement.


## Cancel → Cancellation Flow

Every user who clicks "Cancel" ultimately confirms. The time between is typically under 2 minutes.

In [15]:
# Analyze Cancel → Cancellation Confirmation relationship
cancel_users = set(train[train['page'] == 'Cancel']['userId'].unique())
confirm_users = set(train[train['page'] == 'Cancellation Confirmation']['userId'].unique())

print(f"Users with 'Cancel' page visit: {len(cancel_users)}")
print(f"Users with 'Cancellation Confirmation': {len(confirm_users)}")
print(f"Perfect match: {cancel_users == confirm_users}")

# Time between Cancel and Confirmation
cancel_times = train[train['page'] == 'Cancel'].groupby('userId')['time'].first()
confirm_times = train[train['page'] == 'Cancellation Confirmation'].groupby('userId')['time'].first()

time_diff = (confirm_times - cancel_times).dt.total_seconds()
print(f"\nTime between Cancel and Confirmation:")
print(f"  Median: {time_diff.median():.1f} seconds")
print(f"  Mean: {time_diff.mean():.1f} seconds")
print(f"  Max: {time_diff.max():.1f} seconds ({time_diff.max()/60:.1f} minutes)")

Users with 'Cancel' page visit: 4271
Users with 'Cancellation Confirmation': 4271
Perfect match: True

Time between Cancel and Confirmation:
  Median: 33.0 seconds
  Mean: 89.8 seconds
  Max: 3720.0 seconds (62.0 minutes)


## Behavioral Patterns: Downgrade & Upgrade Events

Downgrade and Upgrade page visits are strong churn indicators. Most churned users went through: Free → Paid (Upgrade) → Active use → Downgrade → Cancel

In [16]:
# Analyze Downgrade/Upgrade patterns for churned vs non-churned users
downgrade_users = set(train[train['page'] == 'Downgrade']['userId'].unique())
upgrade_users = set(train[train['page'] == 'Upgrade']['userId'].unique())

churned_with_downgrade = len(churned_set & downgrade_users)
churned_with_upgrade = len(churned_set & upgrade_users)

print("Churned users behavior:")
print(f"  Had Downgrade event: {churned_with_downgrade:,} ({churned_with_downgrade/len(churned_set):.1%})")
print(f"  Had Upgrade event: {churned_with_upgrade:,} ({churned_with_upgrade/len(churned_set):.1%})")

# Compare with non-churned
non_churned_set = set(all_users) - churned_set
non_churned_with_downgrade = len(non_churned_set & downgrade_users)
non_churned_with_upgrade = len(non_churned_set & upgrade_users)

print(f"\nNon-churned users behavior:")
print(f"  Had Downgrade event: {non_churned_with_downgrade:,} ({non_churned_with_downgrade/len(non_churned_set):.1%})")
print(f"  Had Upgrade event: {non_churned_with_upgrade:,} ({non_churned_with_upgrade/len(non_churned_set):.1%})")

Churned users behavior:
  Had Downgrade event: 3,113 (72.9%)
  Had Upgrade event: 2,981 (69.8%)

Non-churned users behavior:
  Had Downgrade event: 9,073 (61.0%)
  Had Upgrade event: 10,217 (68.7%)


## Subscription Level at Churn Time

What subscription level were users on when they churned?

In [17]:
# Get level at churn time (last record before cancellation)
churned_data = train[train['userId'].isin(churned_set)]
last_level = churned_data.groupby('userId')['level'].last()

print("Subscription level at churn:")
print(last_level.value_counts())
print(f"\nPaid users churning: {(last_level == 'paid').sum()} ({(last_level == 'paid').mean():.1%})")
print(f"Free users churning: {(last_level == 'free').sum()} ({(last_level == 'free').mean():.1%})")

Subscription level at churn:
level
paid    2879
free    1392
Name: count, dtype: int64

Paid users churning: 2879 (67.4%)
Free users churning: 1392 (32.6%)


## Missing Values Explained

The missing values are **structural**, not data quality issues:
- `song/artist/length` are NULL only on non-music pages (Home, Settings, etc.)
- In test set: demographics are NULL for logged-out/guest sessions

In [18]:
# Verify: missing song/artist/length only on non-NextSong pages
nextsong_missing = train[train['page'] == 'NextSong']['song'].isna().sum()
other_pages_missing = train[train['page'] != 'NextSong']['song'].isna().sum()
other_pages_total = len(train[train['page'] != 'NextSong'])

print("Missing song values by page type:")
print(f"  NextSong pages with missing song: {nextsong_missing:,}")
print(f"  Other pages with missing song: {other_pages_missing:,} / {other_pages_total:,} ({other_pages_missing/other_pages_total:.1%})")

# In test: missing demographics are logged-out sessions
print("\nTest set missing demographics by auth status:")
print(test.groupby('auth')['gender'].apply(lambda x: x.isna().sum()))

Missing song values by page type:
  NextSong pages with missing song: 0
  Other pages with missing song: 3,208,203 / 3,208,203 (100.0%)

Test set missing demographics by auth status:
auth
Guest           3363
Logged In          0
Logged Out    650318
Name: gender, dtype: int64


## Train/Test User Overlap

Critical check: are test users in the training set?

In [19]:
# Check user overlap between train and test
train_users = set(train['userId'].unique())
test_users = set(test['userId'].unique())

overlap = train_users & test_users
print(f"Train users: {len(train_users):,}")
print(f"Test users: {len(test_users):,}")
print(f"Overlap: {len(overlap)} users ({len(overlap)/len(test_users):.1%} of test)")

if len(overlap) == 0:
    print("\n✓ No overlap - model must generalize to unseen users")
    print("  → Features must capture behavior patterns, not user identity")

Train users: 19,140
Test users: 2,904
Overlap: 0 users (0.0% of test)

✓ No overlap - model must generalize to unseen users
  → Features must capture behavior patterns, not user identity


## Page Types: Train vs Test Differences

Some pages only appear in train (Cancel, Cancellation Confirmation) or test (Login, Register).

In [20]:
# Compare page types between train and test
train_pages = set(train['page'].unique())
test_pages = set(test['page'].unique())

print("Pages only in TRAIN (churn-related):")
for p in sorted(train_pages - test_pages):
    print(f"  - {p}")

print("\nPages only in TEST (new user flow):")
for p in sorted(test_pages - train_pages):
    print(f"  - {p}")

Pages only in TRAIN (churn-related):
  - Cancel
  - Cancellation Confirmation

Pages only in TEST (new user flow):
  - Login
  - Register
  - Submit Registration


## Gender Distribution in Churned Users

In [21]:
# Gender distribution in churned users
churned_gender = churned_data.groupby('userId')['gender'].first().value_counts()
print("Gender of churned users:")
print(churned_gender)
print(f"\nMale churn: {churned_gender.get('M', 0)/len(churned_set):.1%}")
print(f"Female churn: {churned_gender.get('F', 0)/len(churned_set):.1%}")
print("\n→ Relatively balanced, no strong gender bias in churn")

Gender of churned users:
gender
M    2277
F    1994
Name: count, dtype: int64

Male churn: 53.3%
Female churn: 46.7%

→ Relatively balanced, no strong gender bias in churn


---

## Summary: Key Takeaways for Modeling

| Finding | Implication for Modeling |
|---------|--------------------------|
| 22% churn rate | Moderate imbalance - use `class_weight='balanced'` |
| Downgrade events in 73% of churners (vs 61% non-churners) | Strong predictive feature |
| Cancel → Confirm in ~30 seconds | Exclude Cancel visits within 12h of churn to avoid leakage |
| No train/test user overlap | Must use behavioral features, not user IDs |
| Missing values are structural | No imputation needed for song/artist |
| 67% churners were paid users | Paid → Free → Cancel is common path |